Socket grpc

pascalwhoop opened the pull request:
Added GRPC variant to allow remote system access to broker-adapter.

pascalwhoop commented on broker.properties:
```diff
@@ -22,4 +22,6 @@ samplebroker.core.jmsManagementService.jmsBrokerUrl = tcp://localhost:61616
\ No newline at end of file
+#samplebroker.core.powerTacBroker.interactive = truesamplebroker.core.brokerMessageReceiver.rawXml = true
+samplebroker.core.brokerMessageReceiver.rawXml = true
+samplebroker.core.brokerMessageReceiver.xmlForwardType = SocketAdapter
```
Configuration via broker.properties, linked to [this PR](https://github.com/powertac/powertac-core/pulls)

jecollins replied:
Line 25 of broker.properties seems like you lost a line ending.

pascalwhoop commented on pom.xml:
```diff
@@ -91,21 +119,42 @@
+          </execution>
+        </executions>
+      </plugin>
       <!-- javadoc config -->
```
automatically generates sources on compile

pascalwhoop commented on src/main/java/org/powertac/samplebroker/grpc/GrpcSocketAdapter.java:
```diff
@@ -0,0 +1,101 @@
+  @Override
+  public void activate(int timeslot)
+  {
+
```
Do I still need to send the clients a specific xml that notifies them of the activation? Or is this already in the raw xml?

jecollins replied:
The common.msg.TimeslotComplete message, or <ts-done timeslotIndex="nnn"/> in xml, is used to drive activation. No need for any other message.

jecollins replied:
Sorry, the xml did not survive the markdown conversion. Let's try again:
  `<ts-done timeslotIndex="nnn"/>`.

pascalwhoop commented:
- [This](https://github.com/pascalwhoop/broker-python) is a Python implementation that shows how to connect to the socket.
- There is also a small integration test that connects to the socket once.

pascalwhoop commented on pom.xml:
```diff
@@ -1,122 +1,171 @@
+                <artifactId>protobuf-maven-plugin</artifactId>
+                <version>0.5.0</version>
+                <configuration>
+                    <protocArtifact>com.google.protobuf:protoc:3.5.1-1:exe:${os.detected.classifier}</protocArtifact>
```
https://github.com/grpc/grpc-java
I just took the install guides from their official Readme. The line resolves based on the operating system of the build server. I just tried on Linux machines, in a Docker container based on Archlinux and on Debian. What did you build on?

jecollins replied:
I'm running on Linux, Ubuntu 16.04 LTS. It's possible that Eclipse is the problem here, because it seems to build OK if I just do mvn clean test in a term window.

jecollins replied:
I think I found a [description of the problem](https://github.com/trustin/os-maven-plugin#issues-with-eclipse-m2e-or-other-ides) with Eclipse.

pascalwhoop commented on src/main/java/org/powertac/samplebroker/grpc/ControlEventHandlerService.java:
```diff
@@ -14,7 +14,7 @@
+package org.powertac.samplebroker.grpc;
 
 import io.grpc.stub.StreamObserver;
 import org.powertac.broker.adapter.grpc.Booly;
```
I’ll check this again, I’m abroad for a few days so give me some time. But there’s no one waiting for these changes so I guess it’s ok..

pascalwhoop commented:
Hi. I found some more information on it. Seems indeed to be a eclipse thing.
From the documentation of os-maven-plugin:
```
Issues with Eclipse m2e or other IDEs
If you are using IntelliJ IDEA, you should not have any problem.
If you are using Eclipse, you need to install an additional Eclipse plugin because m2e does not evaluate the extension specified in a pom.xml.Download os-maven-plugin-1.5.0.Final.jarand put it into the <ECLIPSE_HOME>/pluginsdirectory.
```
https://github.com/trustin/os-maven-plugin/blob/master/README.md
On 3 Mar 2018 5:43 PM, John Collins <notifications@github.com> wrote:@jecollins commented on this pull request.



In pom.xml:
> +                                <transformer
+                                        implementation="org.apache.maven.plugins.shade.resource.ManifestResourceTransformer">
+                                    <mainClass>org.powertac.samplebroker.core.BrokerMain</mainClass>
+                                </transformer>
+                            </transformers>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <!--GRPC generate files-->
+            <plugin>
+                <groupId>org.xolstice.maven.plugins</groupId>
+                <artifactId>protobuf-maven-plugin</artifactId>
+                <version>0.5.0</version>
+                <configuration>
+                    <protocArtifact>com.google.protobuf:protoc:3.5.1-1:exe:${os.detected.classifier}</protocArtifact>

I'm running on Linux, Ubuntu 16.04 LTS. It's possible that Eclipse is the problem here, because it seems to build OK if I just do mvn clean test in a term window.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.

jecollins commented:
So is the missing powertac.broker package the only remaining problem? Did you perhaps neglect to commit something?

pascalwhoop commented:
Yes I originally called it like that because it seemed to make sense to call it as such but then I saw that all packages are in samplebroker even if it’s not really a “sample” but rather the core of most brokers developed by the teams. So I moved everything and I think the import has just not been cleaned up.

pascalwhoop commented:
But I think there’s several imports that are messed up like that. Let me look into it tomorrow I can ssh home and get an IDE going right now I’m only on an iPad

pascalwhoop commented:
Alright. It was all working, but the grpc proto file was set to generate the classes to a `broker.adapter.grpc` package instead of the `samplebroker.grpc` package. I changed that in the last commit.

jecollins commented:
I still have three unresolved imports in `grpc.ControlEventHandler.java`: `org.powertac.samplebroker.grpc.Booly, org.powertac.samplebroker.grpc.ServerMessagesStreamGrpc, org.powertac.samplebroker.grpc.XmlMessage`. Why these are imported is a mystery, because they are in the same package and are not referenced in the code. But there's still an unresolved type `ControlEvent`. So it still fails to build.

pascalwhoop commented:
Alright, I cleaned those imports up (I added them via vi in ssh so I had no IDE support telling me which imports are unused) and added the import you mentioned. What really irritates me is the fact that it built all these times on my machine and I am starting to wonder where the differences lie in the way we build things. If there is a ressource that shows how exactly the environment looks like and how the artifacts are build that would be useful for me to avoid such errors in the future.

As a note. I build with the CLI doing a simple `mvn clean package` using:
- maven 3.5.0
- OpenJDK 1.8.0_151

The repo is in the exact state as the PR suggests except the broker.properties file which has different credentials. It’s really beyond me how mine succeeds but yours fails. 

I ran this again in a clean container running the latest Ubuntu LTS.

```
[INFO] Building jar: /broker-adapter/target/broker-adapter-javadoc.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 03:54 min
[INFO] Finished at: 2018-03-07T11:07:18+00:00
[INFO] Final Memory: 32M/394M
[INFO] ------------------------------------------------------------------------
root@4a1295b03483:/broker-adapter# history
    1  sudo apt update
    2  apt update
    3  apt install default-jdk maven
    4  git
    5  apt install -y git
    6  git clone https://github.com/pascalwhoop/broker-adapter
    7  cd broker-adapter/
    8  ls
    9  mvn clean package
   10  history
```

jecollins commented:
That's very strange, because it builds here with maven as well. Disturbing.

In the longer run, I would like to do a bit of refactoring to undo the changes to broker-core, although I haven't yet thought it through. I don't like having to change broker-core to accommodate these kinds of changes. For example, a broker developer should be able to add command-line options or change message-handling without changing broker-core.

jecollins commented:
OK, but the commit still leaves two imports unresolved: `org.powertac.samplebroker.grpc.Booly`, and `org.powertac.samplebroker.grpc.ControlEvent`. It builds, but complains about unchecked or unsafe operations.

jecollins commented:
It seems these issues are being caused by generated code that's not seen by Eclipse. The imported classes are indeed present in target/generated-sources.

jecollins commented:
[Updated pom.xml according to this issue](https://github.com/querydsl/apt-maven-plugin/issues/50) and imports are now resolved in eclipse.

jecollins commented:
Original R broker fails because broker.properties specifies "xmlForwardType=SocketAdapter". I guess we should specify this in the command-line if we want to keep these two versions working in parallel.
